src/custom/state/swap/hooks.ts

```diff
@@ -0,0 +1,123 @@
+import useENS from '@src/hooks/useENS'
+import { Currency, CurrencyAmount, Trade } from '@uniswap/sdk'
+import { useActiveWeb3React } from '@src/hooks'
+import { useCurrency } from '@src/hooks/Tokens'
+import { isAddress } from '@src/utils'
+import { useCurrencyBalances } from '@src/state/wallet/hooks'
+import { Field } from '@src/state/swap/actions'
+import { useUserSlippageTolerance } from '@src/state/user/hooks'
+import { computeSlippageAdjustedAmounts } from '@src/utils/prices'
+import { tryParseAmount, useSwapState } from '@src/state/swap/hooks'
+import { useFee } from '../fee/hooks'
+import { registerOnWindow } from 'utils/misc'
+import { useTradeExactInWithFee, useTradeExactOutWithFee } from './extension'
+
+export * from '@src/state/swap/hooks'
+
+interface DerivedSwapInfo {
+  currencies: { [field in Field]?: Currency }
+  currencyBalances: { [field in Field]?: CurrencyAmount }
+  parsedAmount: CurrencyAmount | undefined
+  v2Trade: Trade | undefined
+  // TODO: review this - we don't use a v1 trade but changing all code
+  // or extending whole swap comp for only removing v1trade is a lot
+  v1Trade: undefined
+  inputError?: string
+}
+
+// from the current swap inputs, compute the best trade and return it.
+export function useDerivedSwapInfo(): DerivedSwapInfo {
+  const { account, chainId } = useActiveWeb3React()
+
+  const {
+    independentField,
+    typedValue,
+    [Field.INPUT]: { currencyId: inputCurrencyId },
+    [Field.OUTPUT]: { currencyId: outputCurrencyId },
+    recipient
+  } = useSwapState()
+
+  const inputCurrency = useCurrency(inputCurrencyId)
+  const outputCurrency = useCurrency(outputCurrencyId)
+  const recipientLookup = useENS(recipient ?? undefined)
+  const to: string | null = (recipient === null ? account : recipientLookup.address) ?? null
```

Comment:
`... ?? null`: doesn't this mean: if left side is `null`, return `null`?

Reply:
this is their code but it is strange. not sure they actually use the coalesced ternary correctly